### PR TITLE
Adds support for Kustomize 2.0.1. Fixes #1085

### LINKED ISSUE #1085: Kustomize 2.0 support
Kustomize 2 is radically different and has many breaking changes, including to the CLI. It's very likely Argo CD will need to simultaneously support both versions for some time, and selectively choose the correct CLI based on the detected version of the kustomization.yaml.

### PR DESCRIPTION
**Pull request by alexec**
This assumes that you have two binaries on your PATH:

* `kustomize` this should be version 1.0.11
* `kustomize2` this should be version 2.0.1

# Testing Locally
If you upgrade brew, you may find you delete one of them to fix-up brew:

* Make sure you have a kustomize-1.0.11.zip from Slack and copy that as /usr/local/Cellar/kustomize/1.0.11
* Make sure you have a kustomize-2.0.1.zip from Slack and copy that as /usr/local/Cellar/kustomize/2.0.1

The make sure you have them on your PATH:

```
cd /usr/local/bin
ln -s ../Cellar/kustomize/1.0.11/bin/kustomize
ln -s ../Cellar/kustomize/2.0.1/bin/kustomize kustomize2
```

You can find a new example app:

https://github.com/alexec/argocd-example-apps/tree/master/kustomize2-guestbook

# Notes on Kustomize 2

This appears to have a new security check, our example code produces this error:

```
Unable to generate manifests in kustomize2-guestbook: rpc error: code = Unknown desc = Error: rawResources failed to read Resources: Load from path ../guestbook/guestbook-ui-deployment.yaml failed: security; file '../guestbook/guestbook-ui-deployment.yaml' is not in or below '/private/var/folders/v5/0lllbd_s0ldg1dr4qxm7vw0scm2ffd/T/https:__github.com_alexec_argocd-example-apps_/kustomize2-guestbook
```

Also, please read https://github.com/kubernetes-sigs/kustomize/blob/master/docs/version2.0.0.md.

### CODE REVIEW

**Comment by jessesuen** [comment]
> This appears to have a new security check, our example code produces this error:

yes this is one of their new security restrictions that I mentioned.

File: manifests/install.yaml [hunk 1]
```diff
@@ -424,7 +424,7 @@ spec:
         - "20"
         - --operation-processors
         - "10"
-        image: argoproj/argocd:latest
+        image: alexcollinsintuit/argocd:latest
```
Comment (jessesuen) [inline manifests/install.yaml]: rerun `make manifests`

File: util/kustomize/kustomize.go [hunk 2]
```diff
@@ -43,15 +59,15 @@ func (k *kustomize) Build(opts KustomizeBuildOpts, overrides []*v1alpha1.Compone
 	}
 
 	for _, override := range overrides {
-		cmd := exec.Command("kustomize", "edit", "set", "imagetag", fmt.Sprintf("%s:%s", override.Name, override.Value))
+		cmd := exec.Command(commandName, "edit", "set", "imagetag", fmt.Sprintf("%s:%s", override.Name, override.Value))
```
Comment (jessesuen) [inline util/kustomize/kustomize.go]: kustomize 2 no longer has `edit set imagetag` and has:
```
$ kustomize-2.0.1 edit set --help
Sets the value of different fields in kustomization file.

Usage:
  kustomize edit set [command]

Examples:

	# Sets the nameprefix field
	kustomize edit set nameprefix <prefix-value>

	# Sets the namesuffix field
	kustomize edit set namesuffix <suffix-value>


Available Commands:
  image       Sets images and their new names, new tags or digests in the kustomization file
  nameprefix  Sets the value of the namePrefix field in the kustomization file.
  namespace   Sets the value of the namespace field in the kustomization file
  namesuffix  Sets the value of the nameSuffix field in the kustomization file.
```

File: util/kustomize/kustomize.go [hunk 3]
```diff
@@ -63,6 +79,61 @@ func (k *kustomize) Build(opts KustomizeBuildOpts, overrides []*v1alpha1.Compone
...
+	if version == 1 {
+		return "kustomize", nil
+	} else {
+		return "kustomize" + strconv.Itoa(version), nil
+	}
+}
+
+// kustomization is a file called kustomization.yaml that describes a configuration consumable by kustomize.
+func (k *kustomize) findKustomization() (string, error) {
+	for _, file := range []string{"kustomization.yaml", "kustomization.yml", "Kustomization"} {
```
Comment (jessesuen) [inline util/kustomize/kustomize.go]: this function will also need to be updated:
* `IdentifyAppSourceTypeByAppDir`
* repository.go::func (s *Server) listAppsPaths()

**Comment by jessesuen** [comment]
As discussed, please make sure kustomize 2 applications returns an empty list for our concept of "parameters". We will file a separate issue to have kustomize specific config in the ApplicationSourceKustomize type. I'll file a separate issue.

File: util/kustomize/kustomize.go (hunk 2 above)
Comment (alexec) [reply util/kustomize/kustomize.go]: I'm actually not sure what we want to do here. Do we want to return an error (i.e. leave it alone?). Do we want to ignore it and log a warning?

**Comment by jessesuen** [comment]
Filed issue https://github.com/argoproj/argo-cd/issues/1141 to track the deprecation of componentParameterOverrides.

File: util/kustomize/kustomize.go [hunk 4]
```diff
@@ -63,6 +79,67 @@ func (k *kustomize) Build(opts KustomizeBuildOpts, overrides []*v1alpha1.Compone
...
+}
+
+func (k *kustomize) getKustomizationVersion() (int, error) {
+
+	kustomization, err := k.findKustomization()
+	if err != nil {
+		return 0, err
+	}
+
+	log.Infof("using kustomization=%s", kustomization)
```
Comment (jessesuen) [inline util/kustomize/kustomize.go]: Minor optimization which won't require opening the file:
```golang
if os.path.Base(kustomization) == "Kustomization" || os.path.Base(kustomization)  == "kustomization.yml" {
    return 2, nil
}
```

File: util/kustomize/kustomize.go [hunk 5]
```diff
@@ -63,6 +79,67 @@ func (k *kustomize) Build(opts KustomizeBuildOpts, overrides []*v1alpha1.Compone
...
+	}
+
+	defer util.Close(file)
+
+	scanner := bufio.NewScanner(file)
+
+	for scanner.Scan() {
+		// +apiVersion: kustomize.config.k8s.io/v1beta1
+		// +kind: Kustomization
+		if strings.Contains(scanner.Text(), "kind: Kustomization") {
```
Comment (jessesuen) [inline util/kustomize/kustomize.go]: This check will be sensitive to extra spaces. You may want to do something like:
```golang
type kustomization struct {
    kind string
}
dat, err := ioutil.ReadFile(kustomization)
if err != nil {
    return 0, err
}
var obj kustomization
err = yaml.Unmarshal(dat, &obj}
if err != nil {
    return 0, err
}
if obj.kind == "Kustomization" {
    return 2, nil
}
return 1, nil
```

File: util/kustomize/kustomize.go (hunk 2 above)
Comment (jessesuen) [reply util/kustomize/kustomize.go]: If we know that this is a kustomize 2 app, we skip the `edit set` command entirely. We will handle this later when we introduce kustomize specific config.

**Review by jessesuen — APPROVED** [review]
I have some minor request. Other than that LGTM

File: util/kustomize/kustomize_test.go [hunk 6]
```diff
@@ -60,3 +65,48 @@ func TestKustomizeBuild(t *testing.T) {
...
+	assert.Equal(t, expected, commandName)
+}
+
+
+func TestIsKustomization(t *testing.T) {
+
+	assert.True(t, IsKustomization("kustomization.yaml"))
+	assert.True(t, IsKustomization("kustomization.yml"))
+	assert.True(t, IsKustomization("Kustomization"))
+	assert.False(t, IsKustomization("rubbish.yml"))
```
[the remark this answers is not in the document]
Comment (alexec) [reply util/kustomize/kustomize_test.go]: Let me add that to my UK to US translation guide.

[the remark this answers is not in the document]
Comment (alexec) [reply util/kustomize/kustomize_test.go]: 🇬🇧

File: server/repository/repository.go [hunk 7]
```diff
@@ -1,11 +1,14 @@
 package repository
 
 import (
+	"errors"
 	"fmt"
 	"path"
 	"path/filepath"
 	"reflect"
 
+	"github.com/argoproj/argo-cd/util/kustomize"
```
Comment (alexmt) [inline server/repository/repository.go]: nitpick: move import down to all argo-cd related imports.

**Review by alexmt — APPROVED** [review]
LGTM. added minor comment

File: server/repository/repository.go (hunk 7 above)
Comment (alexec) [reply server/repository/repository.go]: Can we use `goimports` please?

**Pull request merged**